Type Error in Union

`s=Union(ImageSet(Lambda(n, -sqrt(n*pi/2 - 1 + pi/2)), Integers), ImageSet(Lambda(n, sqrt(n*pi/2 - 1 + pi/2)), Integers))`

`s + FiniteSet(0)` raises TypeError.

Can anybody tell me what is the problem?

In `Intersection.__new__` arg order is canonicalised using `ordered` with the key `Set._infimum_key`. The keys in this case are:
```julia
In [55]: args = [Mul(2, 1/pi, 1-pi/2), Mul(1/pi, (2-pi))]

In [56]: args
Out[56]: 
⎡  ⎛    π⎞       ⎤
⎢2⋅⎜1 - ─⎟       ⎥
⎢  ⎝    2⎠  2 - π⎥
⎢─────────, ─────⎥
⎣    π        π  ⎦

In [57]: sorted(args)
---------------------------------------------------------------------------
TypeError 
```
A key function should return members of a totally ordered set so it is a mistake to try and sort sets by infimum.

I think the fix is this:
```diff
diff --git a/sympy/sets/sets.py b/sympy/sets/sets.py
index 8a50b2b..6f4a6fc 100644
--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -1350,7 +1350,7 @@ def __new__(cls, *args, **kwargs):
             args = list(cls._new_args_filter(args))
             return simplify_intersection(args)
 
-        args = list(ordered(args, Set._infimum_key))
+        args = list(ordered(args))
 
         obj = Basic.__new__(cls, *args)
         obj._argset = frozenset(args)
```
That leads to a couple of failures in solveset:
```
_________________________________ test_nonlinsolve_positive_dimensional _________________________________
sympy/solvers/tests/test_solveset.py:1415: in test_nonlinsolve_positive_dimensional
    assert nonlinsolve(system, [a, b, c, d]) == soln
sympy/solvers/solveset.py:3412: in nonlinsolve
    res = _handle_positive_dimensional(polys, symbols, denominators)
sympy/solvers/solveset.py:3161: in _handle_positive_dimensional
    denominators)
sympy/solvers/solveset.py:3087: in substitution
    old_result, solveset_real)
sympy/solvers/solveset.py:2966: in _solve_using_known_values
    result = _new_order_result(result, eq)
sympy/solvers/solveset.py:2941: in _new_order_result
    if eq.subs(res) == 0:
sympy/core/basic.py:972: in subs
    rv = rv._subs(old, new, **kwargs)
sympy/core/cache.py:96: in wrapper
    retval = cfunc(*args, **kwargs)
sympy/core/basic.py:1086: in _subs
    rv = fallback(self, old, new)
sympy/core/basic.py:1058: in fallback
    arg = arg._subs(old, new, **hints)
sympy/core/cache.py:96: in wrapper
    retval = cfunc(*args, **kwargs)
sympy/core/basic.py:1086: in _subs
    rv = fallback(self, old, new)
sympy/core/basic.py:1063: in fallback
    rv = self.func(*args)
sympy/core/cache.py:96: in wrapper
    retval = cfunc(*args, **kwargs)
sympy/core/operations.py:54: in __new__
    c_part, nc_part, order_symbols = cls.flatten(args)
sympy/core/mul.py:200: in flatten
    r, b = b.as_coeff_Mul()
E   AttributeError: 'Reals' object has no attribute 'as_coeff_Mul'
_____________________________________ test_issue_5132_substitution ______________________________________
sympy/solvers/tests/test_solveset.py:1655: in test_issue_5132_substitution
    assert substitution(system, [x, y]) == soln
sympy/solvers/solveset.py:3087: in substitution
    old_result, solveset_real)
sympy/solvers/solveset.py:2966: in _solve_using_known_values
    result = _new_order_result(result, eq)
sympy/solvers/solveset.py:2941: in _new_order_result
    if eq.subs(res) == 0:
sympy/core/basic.py:972: in subs
    rv = rv._subs(old, new, **kwargs)
sympy/core/cache.py:96: in wrapper
    retval = cfunc(*args, **kwargs)
sympy/core/basic.py:1086: in _subs
    rv = fallback(self, old, new)
sympy/core/basic.py:1058: in fallback
    arg = arg._subs(old, new, **hints)
sympy/core/cache.py:96: in wrapper
    retval = cfunc(*args, **kwargs)
sympy/core/basic.py:1086: in _subs
    rv = fallback(self, old, new)
sympy/core/basic.py:1063: in fallback
    rv = self.func(*args)
sympy/core/cache.py:96: in wrapper
    retval = cfunc(*args, **kwargs)
sympy/core/operations.py:54: in __new__
    c_part, nc_part, order_symbols = cls.flatten(args)
sympy/core/mul.py:200: in flatten
    r, b = b.as_coeff_Mul()
E   AttributeError: 'ProductSet' object has no attribute 'as_coeff_Mul'
```

--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -80,6 +80,7 @@ class Set(Basic):
         try:
             infimum = expr.inf
             assert infimum.is_comparable
+            infimum = infimum.evalf()  # issue #18505
         except (NotImplementedError,
                 AttributeError, AssertionError, ValueError):
             infimum = S.Infinity

--- a/sympy/sets/tests/test_sets.py
+++ b/sympy/sets/tests/test_sets.py
@@ -1437,6 +1437,11 @@ def test_issue_11174():
     assert Intersection(FiniteSet(x), S.Reals) == soln
 
 
+def test_issue_18505():
+    assert ImageSet(Lambda(n, sqrt(pi*n/2 - 1 + pi/2)), S.Integers).contains(0) == \
+            Contains(0, ImageSet(Lambda(n, sqrt(pi*n/2 - 1 + pi/2)), S.Integers))
+
+
 def test_finite_set_intersection():
     # The following should not produce recursion errors
     # Note: some of these are not completely correct. See

